### PR TITLE
Fineract Asset Externalization models

### PR DESCRIPTION
**Pull request by b0c1**
- [x] Liquibase database model
- [x] JPA model
- [x] API endpoint without implementation

### CODE REVIEW

**Review by galovics — APPROVED** [review]

File: fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResource.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    public String transferRequestWithLoanId(@PathParam("loanId") final Long loanId,
+            @QueryParam("command") @Parameter(description = "command") final String commandParam,
+            @Parameter(hidden = true) final String apiRequestBodyAsJson) {
+        platformUserRightsContext.isAuthenticated();
+
+        throw new NotImplementedException("Not implemented yet");
+    }
+
+    @POST
+    @Path("/transfers/loans//external-id/{loanExternalId}")
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/api/ExternalAssetOwnersApiResource.java:57]: double / in the path.

File: fineract-investor/src/main/java/org/apache/fineract/investor/data/ExternalTransferLoanMappingData.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.apache.fineract.investor.data;
+
+import lombok.Data;
+
+@Data
+public class ExternalTransferLoanMappingData {
+
+    private final Long loanId;
+    private final String externalId;
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/data/ExternalTransferLoanMappingData.java:27]: Please use ExternalId class

File: fineract-investor/src/main/java/org/apache/fineract/investor/data/ExternalTransferOwnerData.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * under the License.
+ */
+package org.apache.fineract.investor.data;
+
+import lombok.Data;
+
+@Data
+public class ExternalTransferOwnerData {
+
+    private final String externalId;
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/data/ExternalTransferOwnerData.java:26]: Please use ExternalId class

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/AbstractAuditableCustom.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import javax.persistence.Column;
+import javax.persistence.MappedSuperclass;
+import org.springframework.data.domain.Auditable;
+
+/**
+ * This class is a copy paste from the fineract-provider module since EclipseLink not weaves the class properly if it's
+ * placed in the fineract-core module.
+ */
+@MappedSuperclass
+public abstract class AbstractAuditableCustom extends AbstractPersistableCustom implements Auditable<Long, Long, LocalDateTime> {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/AbstractAuditableCustom.java:32]: Please use `AbstractAuditableWithUTCDateTimeCustom` instead

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/AbstractPersistableCustom.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * This class is a copy paste from the fineract-provider module since EclipseLink not weaves the class properly if it's
+ * placed in the fineract-core module.
+ */
+@MappedSuperclass
+@Getter
+@Setter
+@NoArgsConstructor
+public abstract class AbstractPersistableCustom implements Persistable<Long>, Serializable {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/AbstractPersistableCustom.java:43]: Please use `AbstractAuditableWithUTCDateTimeCustom` instead

Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/AbstractPersistableCustom.java:43]: We have two `AbstractPersistableCustom` now, do we need both of them?

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwner.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Table(name = "m_external_asset_owner")
+public class ExternalAssetOwner extends AbstractAuditableCustom {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwner.java:34]: Please use `AbstractAuditableWithUTCDateTimeCustom` instead

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwner.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Column(name = "external_id", nullable = false, length = 100, unique = true)
+    private String externalId;
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwner.java:37]: Please use `ExternalId` class

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwnerTransfer.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Table(name = "m_external_asset_owner_transfer")
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class ExternalAssetOwnerTransfer extends AbstractAuditableCustom {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwnerTransfer.java:36]: Please use `AbstractAuditableWithUTCDateTimeCustom` instead

File: fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwnerTransferLoanMapping.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Table(name = "m_external_asset_owner_transfer_loan_mapping")
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+public class ExternalAssetOwnerTransferLoanMapping extends AbstractAuditableCustom {
```
Comment (adamsaghy) [inline fineract-investor/src/main/java/org/apache/fineract/investor/domain/ExternalAssetOwnerTransferLoanMapping.java:34]: Please use `AbstractAuditableWithUTCDateTimeCustom` instead

**Pull request merged**